service/history/workflowcache/cache.go

```diff
@@ -95,13 +111,33 @@ func (c *wfCache) getCacheItem(domainID string, workflowID string) *cacheValue {
 
 	value, ok := c.lru.Get(key).(*cacheValue)
 
+	if ok {
+		return value, nil
+	}
+
+	value = &cacheValue{
+		externalRateLimiter: c.externalLimiterFactory.GetLimiter(domainID),
+		internalRateLimiter: c.internalLimiterFactory.GetLimiter(domainID),
+	}
+	// PutIfNotExist is thread safe, and will either return the value that was already in the cache or the value we just created
+	// another thread might have inserted a value between the Get and PutIfNotExist, but that is ok
+	// it should never return an error as we do not use Pin
+	valueInterface, err := c.lru.PutIfNotExist(key, value)
+	value, ok = valueInterface.(*cacheValue)
+
+	// This should never happen, either the value was already in the cache or we just inserted it
 	if !ok {
-		value = &cacheValue{
-			externalRateLimiter: c.externalLimiterFactory.GetLimiter(domainID),
-			internalRateLimiter: c.internalLimiterFactory.GetLimiter(domainID),
-		}
-		c.lru.PutIfNotExist(key, value)
+		return nil, errors.New("Failed to insert new value into cache")
 	}
 
-	return value
+	return value, err
+}
+
+func (c *wfCache) logError(domainID string, workflowID string, err error) {
+	c.logger.Error("Unexpected error from workflow cache",
```

Comment:
not in this PR but we can add metrics to cover cache hit/miss/err counts per domain.

Reply:
Yes, that is the next step